contracts/allocators/OnsenAllocatorV2.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.10;
+import "../libraries/Address.sol";
+
+///  ========== TYPES ==========
+import "../types/BaseAllocator.sol";
+
+///  ========== INTERFACES ==========
+import "../interfaces/IMasterChef.sol";
+import "../interfaces/ISushiBar.sol";
+
+/**
+ *  @notice Contract deploys liquidity from treasury into the Onsen program,
+ *  earning $SUSHI that can be staked and/or deposited into the treasury.
+ */
+contract OnsenAllocatorV2 is BaseAllocator {
+    using SafeERC20 for IERC20;
+
+    /// ========== STATE VARIABLES ==========
+
+    address immutable sushi; // $SUSHI token
+    address immutable xSushi; // $xSUSHI token
+    address immutable masterChef; // Onsen contract
+
+    address immutable treasury; // Olympus Treasury
+
+    /// ========== CONSTRUCTOR ==========
+
+    constructor(
+        address _chef,
+        address _sushi,
+        address _xSushi,
+        address _treasury,
+        AllocatorInitData memory data
+    ) BaseAllocator(data) {
+        require(_chef != address(0));
+        masterChef = _chef;
+        require(_sushi != address(0));
+        sushi = _sushi;
+        require(_xSushi != address(0));
+        xSushi = _xSushi;
+        require(_treasury != address(0));
+        treasury = _treasury;
+
+        /**  @dev approve for safety, sushi is being staked and xsushi is being instantly sent to treasury and that is fine
+         *   but to be absolutely safe this one approve won't hurt
+         */
+        IERC20(sushi).approve(address(extender), type(uint256).max);
+        IERC20(xSushi).approve(address(extender), type(uint256).max);
+    }
+
+    /**
+     * @notice Figure out which LP tokens to be deposited into onsen, then deposit it into onsen pool and stake the sushi tokens that are returned.
+     * 1- Deposit LP token into onsen, if deposit is succesfull we should get in return sushi tokens
+     * 2- Stake the Sushi rewards tokens returned from deposit LP tokens and as rewards form onsen LP stake, we should get in return xSushi tokens
+     * 3- Return the xSushi tokens to treasury
+     * 4- Calculate gains/loss
+     */
+    function _update(uint256 id) internal override returns (uint128 gain, uint128 loss) {
+        uint256 index = tokenIds[id];
+        IERC20 LPtoken = _tokens[index];
+        uint256 balance = LPtoken.balanceOf(address(this));
+
+        /// Deposit LP token into onsen, if deposit succesfull this address should have in return sushi tokens
+        if (balance > 0) {
+            IERC20(LPtoken).approve(masterChef, balance);
```

Comment:
This does not need to be cast to IERC20 again. LPtoken is already of the type IERC20

Reply:
Code updated to avoid casting to IERC20 again when it is already IERC20